README.md
# bookstore

## Day 1

### Installing Datomic

On-premise installation:
https://www.datomic.com/on-prem.html 

To create an account:

https://my.datomic.com/account/create 

Version downloaded: `0.9.6045`

To configure an embedded db:

```
cp config/samples/dev-transactor-template.properties config/dev-transactor.properties 
```

And then change `config/dev-transactor.propertie` to include the license key you got by email.

To run the transactor:

```
bin/transactor config/dev-transactor.properties 
```

Installing Maven with [SDKMan](https://sdkman.io/)

```
curl -s "https://get.sdkman.io" | bash 
sdk install maven 3.6.3 
```

Installing Datomic Pro Peer Library into our local Maven repo (`~/.m2`)

```sh
bin/maven-install
```

Alternatives for installing the Peer Library: https://docs.datomic.com/on-prem/integrating-peer-lib.html

### Links

Rich quoting Heraclitus: https://www.youtube.com/watch?v=E4RarTAZ2AY&feature=youtu.be&t=1192 

More Heraclitus quotes: https://en.wikiquote.org/wiki/Heraclitus#Quotes

Clojure’s approach to Identity and State: https://clojure.org/about/state
> The set of my favorite foods doesn’t change, i.e. if I prefer different foods in the future, that will be a different set.

Datomic Rationale and comparison of architecture with traditional DBs: https://docs.datomic.com/on-prem/rationale.html 
More about Datomic's architecture: https://docs.datomic.com/on-prem/architecture.html

Open source Datomic alternative: https://github.com/juxt/crux 

Datomic Schema: https://docs.datomic.com/on-prem/schema.html 


Datalog language, a subset of Prolog: https://en.wikipedia.org/wiki/Datalog

## Day 2

### Links

Docs on the list and map form of Datomic transactions: https://docs.datomic.com/on-prem/transactions.html

Weird Characters in Clojure (like `#inst`): https://clojure.org/guides/weird_characters#tagged_literals

The five common forms of Clojure keywords (via Kazuki): https://blog.jeaye.com/2017/10/31/clojure-keywords/

Datomic Workshop by it's creators: https://docs.datomic.com/on-prem/day-of-datomic.html

## Day 3

### Links

Datomic Courses (in pt_BR): https://www.alura.com.br/formacao-datomic

Intro to DDD: https://www.amazon.com.br/Domain-Driven-Design-Distilled-Vaughn-Vernon/dp/0134434420
Books on DDD: https://dddcommunity.org/books/

Domain Events: https://microservices.io/patterns/data/domain-event.html

Event Storming (the workshop): https://en.wikipedia.org/wiki/Event_storming
Event Storming (the workshop): https://www.eventstorming.com/

Event Sourcing (the technique): https://microservices.io/patterns/data/event-sourcing.html
Event Sourcing (the technique): https://martinfowler.com/eaaDev/EventSourcing.html

Datomic Console: https://docs.datomic.com/on-prem/console.html

Java UUIDs: https://docs.oracle.com/javase/7/docs/api/java/util/UUID.html
Examples of using UUIDs in Java: https://www.baeldung.com/java-uuid

Identity and Uniqueness in Datomic: https://docs.datomic.com/on-prem/identity.html

Clojure books from The Pragmatic Bookshelf: https://pragprog.com/search/?q=clojure

Why's Poignant Guide to Ruby: https://poignant.guide/
Land of Lisp (a Common Lisp book): http://landoflisp.com/
SICP (code in Scheme): https://mitpress.mit.edu/sites/default/files/sicp/index.html


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2020-07-24
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2020-07-24
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/bookstore/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/bookstore/compare/0.1.0...0.1.1


test/bookstore/core_test.clj
(ns bookstore.core-test
  (:require [clojure.test :refer :all]
            [bookstore.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/bookstore/db.clj
(ns bookstore.db
  (:require [datomic.api :as d]))

; SCHEMA is a datom in Datomic
; entity-id -> 72
; attribute -> 10 :db/ident
; value -> :book/name
; tx-id -> 13194139534312
; add? -> true

; BOOK is a datom in Datomic

; #datom[17592186045418 72 "Clojure for the Brave and True" 13194139534313 true]
; entity-id -> 17592186045418
; attribute -> 72 :book/name
; value -> "Clojure for the Brave and True"
; tx-id -> 13194139534313
; add? -> true

; #datom[17592186045418 73 9.99 13194139534313 true]
; entity-id -> 17592186045418
; attribute -> 73 :book/price
; value -> 9.99
; tx-id -> 13194139534313
; add? -> true

; #datom [entity-id attribute value tx-id add?]


;#object[datomic.promise$settable_future$reify__7382
;        0x468c92cc
;        {:status :ready,
;         :val {:db-before datomic.db.Db,
;               @9dc5beef :db-after,
;               datomic.db.Db @16418ddb,
;               :tx-data [#datom[13194139534317 50 #inst"2020-07-24T14:36:22.331-00:00" 13194139534317 true]
;                         #datom[17592186045422 72 "SICP" 13194139534317 true]
;                         #datom[17592186045422 73 450.5 13194139534317 true]],
;               :tempids {-9223301668109598135 17592186045422}}}]

; :tx-data [#datom[13194139534315 50 #inst"2020-07-24T14:34:00.636-00:00" 13194139534315 true]
;                         #datom[17592186045420 72 "Clojure for the Brave and True" 13194139534315 true]
;                         #datom[17592186045420 73 99.9 13194139534315 true]],

; result of the transact of the schema
;#object[datomic.promise$settable_future$reify__7382
;        0x2fef55f9
;        {:status :ready,
;         :val {:db-before datomic.db.Db,
;               @95ad4c6a :db-after,
;               datomic.db.Db @e944d5d4,
;               :tx-data [#datom[13194139534312 50 #inst"2020-07-24T13:50:28.777-00:00" 13194139534312 true]
;                         #datom[72 10 :book/name 13194139534312 true]
;                         #datom[72 40 23 13194139534312 true]
;                         #datom[72 41 35 13194139534312 true]
;                         #datom[72 62 "The name of the book" 13194139534312 true]
;                         #datom[73 10 :book/price 13194139534312 true]
;                         #datom[73 40 57 13194139534312 true]
;                         #datom[73 41 35 13194139534312 true]
;                         #datom[73 62 "The price of the book" 13194139534312 true]
;                         #datom[0 13 72 13194139534312 true]
;                         #datom[0 13 73 13194139534312 true]],
;               :tempids {-9223301668109598139 72, -9223301668109598138 73}}}]
;(d/transact conn [clojure-book])
;=>
;#object[datomic.promise$settable_future$reify__7382
;        0x646ccaad
;        {:status :ready,
;         :val {:db-before datomic.db.Db,
;               @afebdb5b :db-after,
;               datomic.db.Db @ee2de7dc,
;               :tx-data [#datom[13194139534313 50 #inst"2020-07-24T13:51:18.739-00:00" 13194139534313 true]
;                         #datom[17592186045418 72 "Clojure for the Brave and True" 13194139534313 true]
;                         #datom[17592186045418 73 9.99 13194139534313 true]],
;               :tempids {-9223301668109598137 17592186045418}}}]


(def db-uri "datomic:dev://localhost:4334/bookstore")

(def schema [
             ; a book
             {:db/ident       :book/id
              :db/valueType   :db.type/uuid
              :db/cardinality :db.cardinality/one
              :db/doc         "The UUID of the book"
              :db/unique      :db.unique/identity}
             {:db/ident       :book/name
              :db/valueType   :db.type/string
              :db/cardinality :db.cardinality/one
              :db/doc         "The name of the book"}
             {:db/ident       :book/price
              :db/valueType   :db.type/double
              :db/cardinality :db.cardinality/one
              :db/doc         "The price of the book"}
             {:db/ident       :book/categories
              :db/valueType   :db.type/string
              :db/cardinality :db.cardinality/many
              :db/doc         "The categories of the book"}
             {:db/ident :book/publisher
              :db/valueType :db.type/ref
              :db/cardinality :db.cardinality/one
              :db/doc "The publisher of the book"}

             ; a publisher
             {:db/ident :publisher/id
              :db/valueType :db.type/uuid
              :db/cardinality :db.cardinality/one
              :db/unique :db.unique/identity}
             {:db/ident :publisher/name
              :db/valueType :db.type/string
              :db/cardinality :db.cardinality/one}])

(defn create-schema! [conn]
  (d/transact conn schema))

(defn open-connection! []
  (d/create-database db-uri)
  (d/connect db-uri))

(defn delete-db! []
 (d/delete-database db-uri))

(defn add-book! [conn book]
  (d/transact conn [book]))

(defn add-publisher! [conn publisher]
  (d/transact conn [publisher]))

(defn add-books! [conn books]
  (d/transact conn books))

(defn update-book! [conn entity-id attribute value]
  (d/transact conn [[:db/add entity-id attribute value]]))

;(d/transact conn [[:db/retract 17592186045425 :book/price]])

; won't work
; transact with the list form (I need the lookup ref)
;(d/transact conn [[:db/add
;                   #uuid"b1f4cf1b-d0d5-4faf-b7e1-ea2c2fbc5708"
;                   :book/name
;                   "Getting Clojure"]])

;doesn't need the lookup ref for the map form
;(d/transact conn [{:book/id #uuid"b1f4cf1b-d0d5-4faf-b7e1-ea2c2fbc5708"
;                   :book/name "Getting Clojure"}])


; datomic.db.Db@16418ddb (it's some moment for this DB)
;(d/db conn)

;(defn find-sicp [conn]
;   (d/q '[:find ?e
;         :where [?e :book/name "SICP"]] (d/db conn)))

;(defn find-books-that-cost-99.9 [conn]
;  (d/q '[:find ?e
;         :where [?e :book/price 99.9]] (d/db conn)))

(defn find-all-entity-ids-of-books [conn]
  (d/q '[:find ?e
         :in $                                              ; implicit
         :where [?e :book/price]]
       (d/db conn)))

(defn find-entity-ids-by-name [conn name-parameter]
  (d/q '[:find ?entity-id
         :in $ ?name                                        ; need to pass db ($) explicitly
         :where [?entity-id :book/name ?name]]
       (d/db conn)
       name-parameter))

(defn find-entity-ids-by-price [conn price-parameter]
  (d/q '[:find ?e
         :in $ ?price
         :where [?e :book/price ?price]]
       (d/db conn)
       price-parameter))

(defn find-all-prices [conn]
  (d/q '[:find ?price
         :where [_ :book/price ?price]]
       (d/db conn)))

(defn find-all-names [conn]
  (d/q '[:find ?name
         :where [_ :book/name ?name]]
       (d/db conn)))

(defn find-all-price-with-entity-ids [conn]
  (d/q '[:find ?e ?price
         :where [?e :book/price ?price]]
       (d/db conn)))

(defn find-all-names-with-entity-ids [conn]
  (d/q '[:find ?e ?name
         :where [?e :book/name ?name]]
       (d/db conn)))

(defn find-all-prices-and-names-with-entity-ids [conn]
  (d/q '[:find ?e ?name ?price
         :where [?e :book/name ?name]
                [?e :book/price ?price]]
       (d/db conn)))

;this will get a Cross Join
;[_ :book/name ?name]
;[_ :book/price ?price]
(defn find-all-prices-and-names [conn]
  (d/q '[:find ?name ?price
         :keys :book/name :book/price
         :where [?e :book/name ?name]
                [?e :book/price ?price]]
       (d/db conn)))

(defn find-all-books [db]
  (d/q '[:find (pull ?e [*])
         :where [?e :book/name ]]
       db))

(defn find-all-books-that-cost-more-than [db minimum-price]
  (d/q '[:find (pull ?e [*])
         :in $ ?min-price
         :where [?e :book/price ?price]
                [(>= ?price ?min-price)]]
       db minimum-price))

(defn find-all-books-by-category [db searched-category]
  (d/q '[:find (pull ?e [*])
         :in $ ?category
         :where [?e :book/categories ?category]]
       db searched-category))

(defn find-book-by-id [db book-id]
  (d/pull db '[*] book-id))

(defn find-all-publishers [db]
  (d/q '[:find (pull ?e [*])
         :where [?e :publisher/id]]
       db))

;(defn find-all-books-that-cost-more-than [db minimum-price]
;  (d/q '[:find ?e ?name ?price
;         :in $ ?min-price
;         :keys :db/id :book/name :book/price
;         :where [?e :book/price ?price]
;         [(>= ?price ?min-price)]
;         [?e :book/name ?name]]
;       db minimum-price))

; buy more books from the publisher
; query price > 200 & quantity < 5

; 10k books
;   - 5k have prices > 200
;   - just 10 have quantity < 5

; -> filtering 5k and then getting 10 => only 3 books
; -> filtering 10 and then getting cost > 200 => only 3 books

; data clause in :where (some thing are "open")
; [entity-id attribute value tx-id added?]


src/bookstore/core.clj
(ns bookstore.core
  (:use clojure.pprint)
  (:require [datomic.api :as d]
            [bookstore.db :as b.db]
            [bookstore.domain :as b.domain]))

(def conn (b.db/open-connection!))
;(println conn)

(b.db/create-schema! conn)

;(b.db/add-book! conn clojure-book)
;(b.db/add-book! conn sicp-book)

(let [clojure-book {:book/name  "Clojure for the Brave and True"
                    :book/price 99.9}
      sicp-book {:book/name "SICP"
                 :book/price 450.5}]
  (b.db/add-books! conn [clojure-book sicp-book]))

(b.db/find-entity-ids-by-name conn "SICP")
(b.db/find-entity-ids-by-name conn "Clojure for the Brave and True")
(b.db/find-entity-ids-by-name conn "Book that doesn't exist")

(b.db/find-entity-ids-by-price conn 99.9)
(b.db/find-entity-ids-by-price conn 450.5)
(b.db/find-entity-ids-by-price conn 123.4)

(b.db/find-all-entity-ids-of-books conn)

(let [kafka-book {:book/name "Metamorphosis"}]
  (b.db/add-book! conn kafka-book))

; didn't get transacted into datomic
; :db.error/nil-value Nil is not a legal value
;(let [stu-book {:book/name "Programming Clojure"
;                :book/price nil}]
;  (b.db/add-book! conn stu-book))

(b.db/find-entity-ids-by-name conn "Metamorphosis")

(b.db/update-book! conn 17592186045421 :book/price 10.0)
;[:db/add entity-id attribute value]

(b.db/find-entity-ids-by-price conn 10.0)

;Promotion!
(b.db/update-book! conn 17592186045419 :book/price 225.25)

(b.db/find-entity-ids-by-price conn 450.5)
(b.db/find-entity-ids-by-price conn 225.25)

(let [stu-book {:book/name "Programming Clojure"
                :book/price 1000.0}]
  (b.db/add-book! conn stu-book))

(b.db/find-entity-ids-by-price conn 1000.0)

(b.db/find-all-prices conn)
;#{[99.9] [225.25] [10.0]}
;Clojure / SICP / Metamorphosis

(b.db/find-all-names conn)
;#{["Programming Clojure"] ["SICP"] ["Clojure for the Brave and True"] ["Metamorphosis"]}

(b.db/find-entity-ids-by-name conn "Programming Clojure")

(b.db/update-book! conn 17592186045425 :book/price 99.9)

(b.db/find-all-prices conn)
;#{[99.9] [225.25] [10.0]}

(b.db/find-all-price-with-entity-ids conn)

(b.db/find-all-names-with-entity-ids conn)

(b.db/find-all-prices-and-names-with-entity-ids conn)

(b.db/find-all-prices-and-names conn)
;#{["SICP" 225.25] ["Clojure for the Brave and True" 99.9] ["Programming Clojure" 99.9] ["Metamorphosis" 10.0]}

(b.db/find-all-books (d/db conn))
;=>
;[[{:db/id 17592186045425, :book/name "Programming Clojure", :book/price 99.9}]
; [{:db/id 17592186045418, :book/name "Clojure for the Brave and True", :book/price 99.9}]
; [{:db/id 17592186045419, :book/name "SICP", :book/price 225.25}]
; [{:db/id 17592186045421, :book/name "Metamorphosis", :book/price 10.0}]]

(def picture-from-the-past (d/db conn))
; => datomic.db.Db@398c5a7f

(let [joy-book {:book/name  "Joy of Clojure"
                    :book/price 400.0}
      applied-book {:book/name "Clojure Applied"
                 :book/price 450.5}]
  (b.db/add-books! conn [joy-book applied-book]))
; :db-before datomic.db.Db, @398c5a7f
; :db-after, datomic.db.Db @9ee30a36,


(def picture-from-now (d/db conn))

(b.db/find-all-books (d/db conn))
(b.db/find-all-books picture-from-the-past)
(b.db/find-all-books picture-from-now)

(b.db/find-all-books (d/as-of (d/db conn)
                              #inst "2014-05-19T19:12:37.925-00:00"))
(b.db/find-all-books (d/as-of (d/db conn)
                              #inst"2020-07-28T12:08:46.350-00:00"))
(b.db/find-all-books (d/as-of (d/db conn)
                              #inst"2020-07-27T14:58:57.789-00:00"))
(b.db/find-all-books (d/as-of (d/db conn)
                              #inst "2024-05-19T19:12:37.925-00:00"))

(b.db/find-all-books-that-cost-more-than (d/db conn) 200.0)

(b.db/create-schema! conn)

(let [whys-book {:book/name "Why's Poignant Guide to Ruby"
                 :book/price 100.0
                 :book/categories ["technical" "ruby"]}]
  (b.db/add-books! conn [whys-book]))

(b.db/find-all-books (d/db conn))

;won't work
;Value [\"technical\" \"clojure\"] is not a valid :string for attribute :book/categories
;(b.db/update-book!
;  conn 17592186045425 :book/categories ["technical" "clojure"])

(b.db/update-book! conn 17592186045425 :book/categories "technical")
(b.db/update-book! conn 17592186045425 :book/categories "clojure")

(b.db/update-book! conn 17592186045428 :book/categories "technical")
(b.db/update-book! conn 17592186045428 :book/categories "clojure")

(b.db/update-book! conn 17592186045429 :book/categories "technical")
(b.db/update-book! conn 17592186045429 :book/categories "clojure")

(b.db/update-book! conn 17592186045418 :book/categories "technical")
(b.db/update-book! conn 17592186045418 :book/categories "clojure")

(b.db/update-book! conn 17592186045419 :book/categories "technical")
(b.db/update-book! conn 17592186045419 :book/categories "scheme")

(b.db/update-book! conn 17592186045421 :book/categories "non-technical")

(b.db/find-all-books (d/db conn))

(b.db/find-all-books-by-category (d/db conn) "clojure")
(b.db/find-all-books-by-category (d/db conn) "non-technical")

(b.db/find-book-by-id (d/db conn) 17592186045419)

(b.db/add-book! conn
                (b.domain/new-book "Computer Networking" 650.5))

(b.db/add-book! conn
                (b.domain/new-book (b.domain/uuid) "Fluent Python" 150.5))

; a db.unique/identity represents a identity
(b.db/add-book! conn
                (b.domain/new-book #uuid"dcf879f7-323e-41f2-be46-8cb723e0d87e"
                                   "TDD by Example" 50.5))

(b.db/find-all-books (d/db conn))

(b.db/add-publisher! conn (b.domain/new-publisher "PragProg"))
(b.db/add-publisher! conn (b.domain/new-publisher "Manning"))
(b.db/add-publisher! conn (b.domain/new-publisher "O'Reilly"))
(b.db/add-publisher! conn (b.domain/new-publisher "No Starch"))
(b.db/add-publisher! conn (b.domain/new-publisher "MIT Press"))

(b.db/find-all-publishers (d/db conn))

; book -> publisher
; 17592186045418 -> 17592186045458

(b.db/update-book! conn 17592186045418 :book/publisher 17592186045458)

(b.db/add-book! conn (b.domain/new-book "Getting Clojure" 125.95))

;#uuid"b1f4cf1b-d0d5-4faf-b7e1-ea2c2fbc5708" -> #uuid"27b43fd6-e446-4dcc-9635-37cdbf7785c2"

;won't work
;Unable to resolve entity: b1f4cf1b-d0d5-4faf-b7e1-ea2c2fbc5708 in datom [#uuid \"b1f4cf1b-d0d5-4faf-b7e1-ea2c2fbc5708\" :book/publisher #uuid \"27b43fd6-e446-4dcc-9635-37cdbf7785c2\"]
;(b.db/update-book! conn
;                   #uuid"b1f4cf1b-d0d5-4faf-b7e1-ea2c2fbc5708"
;                   :book/publisher
;                   #uuid"27b43fd6-e446-4dcc-9635-37cdbf7785c2")

; needs a Lookup Ref
(b.db/update-book! conn
                   [:book/id #uuid"b1f4cf1b-d0d5-4faf-b7e1-ea2c2fbc5708"]
                   :book/publisher
                   [:publisher/id #uuid"27b43fd6-e446-4dcc-9635-37cdbf7785c2"])


(b.db/add-book! conn (b.domain/new-book "Land of Lisp" 165.80))
;#uuid"436f262f-56fb-4a86-9401-4b612eaf330a"

(d/transact conn [{:book/id        #uuid"436f262f-56fb-4a86-9401-4b612eaf330a"
                   :book/publisher [:publisher/id #uuid"fb8f16d9-7024-4b65-a095-9fdccf7e787a"]}])

;(b.db/delete-db!)
